core-bundle/src/Security/Voter/DataContainer/PagePermissionVoter.php

```diff
@@ -0,0 +1,233 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Security\Voter\DataContainer;
+
+use Contao\BackendUser;
+use Contao\CoreBundle\Framework\ContaoFramework;
+use Contao\CoreBundle\Security\ContaoCorePermissions;
+use Contao\CoreBundle\Security\DataContainer\CreateAction;
+use Contao\CoreBundle\Security\DataContainer\DeleteAction;
+use Contao\CoreBundle\Security\DataContainer\ReadAction;
+use Contao\CoreBundle\Security\DataContainer\UpdateAction;
+use Contao\Database;
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+use Symfony\Component\Security\Core\Authorization\AccessDecisionManagerInterface;
+use Symfony\Component\Security\Core\Authorization\Voter\CacheableVoterInterface;
+use Symfony\Component\Security\Core\Authorization\Voter\VoterInterface;
+use Symfony\Contracts\Service\ResetInterface;
+
+/**
+ * @internal
+ */
+class PagePermissionVoter implements VoterInterface, CacheableVoterInterface, ResetInterface
+{
+    private array $pagemountsCache = [];
+
+    public function __construct(
+        private readonly ContaoFramework $framework,
+        private readonly AccessDecisionManagerInterface $accessDecisionManager,
+    ) {
+    }
+
+    public function supportsAttribute(string $attribute): bool
+    {
+        return \in_array($attribute, [ContaoCorePermissions::DC_PREFIX.'tl_page', ContaoCorePermissions::DC_PREFIX.'tl_article'], true);
+    }
+
+    public function supportsType(string $subjectType): bool
+    {
+        return \in_array($subjectType, [CreateAction::class, ReadAction::class, UpdateAction::class, DeleteAction::class], true);
+    }
+
+    public function reset(): void
+    {
+        $this->pagemountsCache = [];
+    }
+
+    public function vote(TokenInterface $token, $subject, array $attributes): int
+    {
+        foreach ($attributes as $attribute) {
+            if (!$this->supportsAttribute($attribute)) {
+                continue;
+            }
```

Comment:
This code skips the loop if the attribute is not supported, but `$attribute` is never used in the rest of the `foreach` loop.

Reply:
not sure what you're referring to. The attribute tells the voter what to vote on, if it is not supported we must not vote.

Reply:
But where exactly do we vote on the attribute? We only vote on `$subject` in the following lines. `$attribute` does not occur after line 60 anymore.

Reply:
we do not vote if the attribute is not supported. In a regular voter, that means to `ABSTAIN`. Just because our voters are _special_ in a sense that they return `ABSTAIN` to `ALLOW` does not mean we should not skip attribute we do not support?

Reply:
This is the same problem as with the `foreach` loops above.